### PR TITLE
arch/risc-v/mpfs: Remove big kernel lock from several drivers

### PR DESCRIPTION
**Pull request by pussuw**
## Summary

This replaces the big kernel lock (enter/leave_critical_section()) with local spinlocks from all MPFS drivers that still use it. The reason is obviously to alleviate big kernel lock congestion which has a massive performance impact in SMP

## Impact

Only the MPFS target is impacted.
App impact: No
Doc impact: No
Build system impact: No

## Testing

Tested with private downstream MPFS target that uses all of the changed drivers.

### CODE REVIEW

**Comment by nuttxpr** [comment]
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13552&type=issues)



The PR mostly meets the NuttX requirements but is missing some crucial information. Here's a breakdown:

**Strengths:**

* **Clear Summary of "Why" and "What":**  The summary clearly explains the purpose of the change (performance improvement by replacing the big kernel lock) and which part of the code is affected (MPFS drivers).
* **Testing Confirmation:** The author confirms testing, which is essential.
* **Concise Impact Assessment:** The impact section is concise and addresses several key areas.

**Weaknesses and Missing Information:**

* **Insufficient "How" in Summary:** The summary lacks detail on *how* the change is implemented.  It mentions using local spinlocks, but doesn't describe the implementation strategy.  Did they create new spinlocks for each driver?  How are they integrated? This detail is critical for reviewers.
* **Lack of Issue References:**  There's no mention of related NuttX or NuttX Apps issues.  If this addresses a known issue, those links are important.
* **Vague Testing Description:** "Tested with private downstream MPFS target" is too vague.  Reviewers need more specifics:
    * Which MPFS target? Provide the configuration.
    * What tests were run?  Just stating "it works as intended" is insufficient. Describe the specific tests and their results.  How was performance measured and what were the improvements?
    * "Private downstream" is a red flag.  Ideally, tests should be reproducible by others. Consider how to make the test setup more accessible (e.g., providing configuration details or contributing the downstream target back upstream if possible).
* **Missing "Before/After" Logs:** The testing logs sections are empty.  These are crucial for demonstrating the impact of the change and verifying its correctness. Include relevant log snippets or performance measurements.
* **Incomplete Impact Assessment:**
    * **Hardware Impact:** While the PR says only the MPFS target is impacted, it should specify which architectures/boards were tested. Changes to locking mechanisms can have subtle architecture-specific effects.
    * **Security Impact:**  Changes to locking could have security implications (e.g., introducing new race conditions).  Even if the impact is "NO," it's important to explicitly state that the security implications were considered.
    * **Compatibility Impact:** Similarly, backward compatibility should be explicitly addressed, even if the answer is "NO."
* **Missing Build Host Information:** The testing section doesn't describe the build host environment used.


**Recommendation:**

The PR needs revisions to address the missing information and improve the level of detail.  Specifically, the author should:

1. **Expand the "How" in the summary:** Describe the spinlock implementation strategy.
2. **Link related issues:** If applicable.
3. **Provide detailed testing information:** Specify the target, configuration, tests performed, and performance results. Make the test setup reproducible if possible.
4. **Include "Before/After" logs:**  Demonstrate the change's impact with concrete data.
5. **Complete the impact assessment:** Provide specific information about hardware tested, security considerations, and backward compatibility.
6. **Specify the build host environment:** OS, CPU, compiler.


By addressing these points, the PR will be much more complete and easier for reviewers to evaluate.

**Review by jerpelea — APPROVED** [review]

**Review by lupyuen — APPROVED** [review]

**Review by acassis — APPROVED** [review]

**Comment by acassis** [comment]
@pussuw nice work! Is there some way to test and display how much it improved the NuttX performance?

**Comment by pussuw** [comment]
> @pussuw nice work! Is there some way to test and display how much it improved the NuttX performance?

I cannot share logs but in our use case total CPU load drops by ~3-5% in SMP mode. This is of course very situational.

The big kernel lock is very contested because semaphores use it so removing its use from modules that don't need it usually has a noticeable impact.

File: arch/risc-v/src/mpfs/mpfs_dsn.c [hunk 1]
```diff
@@ -91,9 +98,7 @@ int mpfs_read_dsn(uint8_t *dsn, size_t len)
 
   while ((getreg32(SERVICES_SR) & SCBCTRL_SERVICESSR_BUSY) && --retries > 0)
     {
-      leave_critical_section(flags);
       nxsig_usleep(1000);
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/mpfs/mpfs_dsn.c:101]: it's wrong to sleep while holding spinlock

File: arch/risc-v/src/mpfs/mpfs_coremmc.c [hunk 2]
```diff
@@ -525,7 +526,7 @@ static void mpfs_configxfrints(struct mpfs_dev_s *priv, uint32_t xfrmask,
 {
   irqstate_t flags;
 
-  flags = enter_critical_section();
+  flags = spin_lock_irqsave(&priv->lock);
   priv->xfrmask = xfrmask;
   priv->xfr_blkmask = xfr_blkmask;
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/mpfs/mpfs_coremmc.c:532]: mcinfo may sleep

File: arch/risc-v/src/mpfs/mpfs_coremmc.c [hunk 3]
```diff
@@ -2411,5 +2412,5 @@ void mpfs_coremmc_sdio_wrprotect(struct sdio_dev_s *dev, bool wrprotect)
 
   mcinfo("cdstatus: %02" PRIx8 "\n", priv->cdstatus);
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/mpfs/mpfs_coremmc.c:2413]: ditto

File: arch/risc-v/src/mpfs/mpfs_emmcsd.c [hunk 4]
```diff
@@ -644,14 +645,14 @@ static void mpfs_configxfrints(struct mpfs_dev_s *priv, uint32_t xfrmask)
 {
   irqstate_t flags;
 
-  flags = enter_critical_section();
+  flags = spin_lock_irqsave(&priv->lock);
   priv->xfrmask = xfrmask;
 
   mcinfo("Mask: %08" PRIx32 "\n", priv->xfrmask | priv->waitmask);
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/mpfs/mpfs_emmcsd.c:651]: ditto

File: arch/risc-v/src/mpfs/mpfs_emmcsd.c [hunk 5]
```diff
@@ -3112,5 +3109,5 @@ void mpfs_emmcsd_sdio_wrprotect(struct sdio_dev_s *dev, bool wrprotect)
 
   mcinfo("cdstatus: %02" PRIx8 "\n", priv->cdstatus);
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/mpfs/mpfs_emmcsd.c:3110]: ditto

File: arch/risc-v/src/mpfs/mpfs_usb.c [hunk 6]
```diff
@@ -477,7 +478,7 @@ static void mpfs_disableclk(void)
                   0);
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/mpfs/mpfs_usb.c:478]: ditto

File: arch/risc-v/src/mpfs/mpfs_coremmc.c (hunk 2 above)
Comment (pussuw) [reply arch/risc-v/src/mpfs/mpfs_coremmc.c:532]: Should the trace be removed or?

File: arch/risc-v/src/mpfs/mpfs_dsn.c (hunk 1 above)
Comment (pussuw) [reply arch/risc-v/src/mpfs/mpfs_dsn.c:101]: The old implementation is equally wrong then, I think we need mutex here instead?

Comment (pussuw) [reply arch/risc-v/src/mpfs/mpfs_dsn.c:101]: Or maybe just remove the loop.. the device should be free if the lock is available

File: arch/risc-v/src/mpfs/mpfs_coremmc.c (hunk 2 above)
Comment (xiaoxiang781216) [reply arch/risc-v/src/mpfs/mpfs_coremmc.c:532]: yes, it's simple fix.

File: arch/risc-v/src/mpfs/mpfs_dsn.c (hunk 1 above)
Comment (xiaoxiang781216) [reply arch/risc-v/src/mpfs/mpfs_dsn.c:101]: could be

Comment (pussuw) [reply arch/risc-v/src/mpfs/mpfs_dsn.c:102]: All is now fixed, thanks for the comments

**Review by jerpelea — APPROVED** [review]

**Review by xiaoxiang781216 — APPROVED** [review]
